value.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"sync"
+
+	promproto "github.com/prometheus/client_model/go"
+	"go.uber.org/atomic"
+)
+
+// Value is an atomic with some associated metadata. It's a building block
+// for higher-level metric types.
+type value struct {
+	atomic.Int64
+
+	meta       metadata
+	labelPairs []*promproto.LabelPair
+}
+
+func newValue(m metadata) value {
+	return value{
+		meta:       m,
+		labelPairs: m.MergeLabels(nil /* variable labels */),
+	}
+}
+
+func newDynamicValue(m metadata, variableLabels []string) value {
+	return value{
+		meta:       m,
+		labelPairs: m.MergeLabels(variableLabels),
+	}
+}
+
+func (v value) snapshot() SimpleSnapshot {
+	return SimpleSnapshot{
+		Name:   *v.meta.Name,
+		Labels: zip(v.labelPairs),
+		Value:  v.Load(),
+	}
+}
+
+// A vector is a collection of values that share the same metadata.
+type vector struct {
+	meta metadata
+
+	// The factory function creates a new metric given the vector's metadata
+	// and the variable label keys and values.
+	factory func(metadata, []string) metric
+
+	metricsMu sync.RWMutex
+	metrics   map[string]metric // key is variable label vals
+}
+
+func (vec *vector) getOrCreate(labels ...string) (metric, error) {
+	if err := vec.meta.ValidateVariableLabels(labels); err != nil {
```

Comment:
admittedly a bit nervous about the Validate logic being distant/detached from the merging logic (which will potentially panic if the validation failed).  Not the biggest deal, but something to be aware of.

Reply:
Yep - that's intentional, since we often merge labels in perf-sensitive situations, but only validate them on metric creation. Agreed that it's touchy, and we'll need to be careful.